chasm/lib/worker/statemachine.go

```diff
@@ -0,0 +1,91 @@
+// State transition logic for Worker.
+package worker
+
+import (
+	"time"
+
+	"go.temporal.io/server/chasm"
+	workerstatepb "go.temporal.io/server/chasm/lib/worker/gen/workerpb/v1"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// updateWorkerLease is a shared helper that updates the worker's lease and schedules expiry.
+func updateWorkerLease(ctx chasm.MutableContext, w *Worker, leaseDeadline time.Time) {
+	w.LeaseExpirationTime = timestamppb.New(leaseDeadline)
+	scheduleLeaseExpiry(ctx, w, leaseDeadline)
+}
+
+// scheduleLeaseExpiry schedules a timer task that will fire when the lease expires.
+func scheduleLeaseExpiry(ctx chasm.MutableContext, w *Worker, leaseDeadline time.Time) {
+	expiryTask := &workerstatepb.LeaseExpiryTask{}
+
+	taskAttrs := chasm.TaskAttributes{
+		ScheduledTime: leaseDeadline,
+		Destination:   "", // Empty means local execution.
+	}
+
+	ctx.AddTask(w, taskAttrs, expiryTask)
+}
+
+// EventHeartbeatReceived is triggered when a heartbeat is received from the worker.
+type EventHeartbeatReceived struct {
+	Time          time.Time
```

Comment:
Don't put time on the events, use `ctx.Now(w)` to get the time from the framework. This is critical to allow us to implement time skipping in the framework later.

Reply:
I removed time fields from other events also. My intent was for future bookkeeping. But realized that chasm can record this automatically when a transition is created or executed.